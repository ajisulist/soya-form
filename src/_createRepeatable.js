import React from 'react';
import PropTypes from 'prop-types';
import { formShape } from './_propTypes';
import {
  createGetName,
  getDisplayName,
} from './_utils';

export default (Component) => {
  class CreateRepeatable extends React.Component {
    static displayName = getDisplayName('CreateRepeatable', Component);

    static propTypes = {
      form: formShape.isRequired,
      length: PropTypes.number.isRequired,
      maxLength: PropTypes.number,
      minLength: PropTypes.number,
      name: PropTypes.arrayOf(PropTypes.string).isRequired,
      addListItem: PropTypes.func.isRequired,
      removeListItem: PropTypes.func.isRequired,
      reorderListItem: PropTypes.func.isRequired,
      reorderListItemDec: PropTypes.func.isRequired,
      reorderListItemInc: PropTypes.func.isRequired,
    };

    static defaultProps = {
      minLength: 0,
    };

    addListItem = () => this.props.addListItem(this.props.name)();

    removeListItem = (index) => this.props.removeListItem(this.props.name)(index);

    reorderListItem = (index, targetIndex) => this.props.reorderListItem(this.props.name)(index, targetIndex);

    reorderListItemDec = (index, amount) => this.props.reorderListItemDec(this.props.name)(index, amount);

    reorderListItemInc = (index, amount) => this.props.reorderListItemInc(this.props.name)(index, amount);

    render() {
      return (
        <div>
          {new Array(this.props.length).fill(null).map((_, index) => (
            <Component
              {...this.props}
              index={index}
              key={index}
              name={createGetName(this.props.name, index)}
              totalItem={this.props.length}
              addListItem={this.addListItem}
              removeListItem={this.removeListItem}
              reorderListItem={this.reorderListItem}
              reorderListItemDec={this.reorderListItemDec}
              reorderListItemInc={this.reorderListItemInc}
            />
          ))}
        </div>
      );
    }
  }

  return CreateRepeatable;
};
