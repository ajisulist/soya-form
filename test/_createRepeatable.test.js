import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import _createRepeatable from '../src/_createRepeatable';

describe('repeatable container', () => {
  const reduceMockProps = (props, name) => ({
    ...props,
    [name]: jest.fn(),
  });

  class Repeatable extends React.Component {
    render() {
      return <div />;
    }
  }

  it('should create 3 repeatable components', () => {
    const RepeatableContainer = _createRepeatable(Repeatable);
    const props = {
      init: jest.fn(),
      length: 3,
      name: ['foo'],
      form: {
        ...[
          'lockSubmission',
          'unlockSubmission',
          'disable',
          'enable',
          'disableField',
          'enableField',
          'disableFields',
          'enableFields',
          'setValue',
          'setValues',
          'setDefaultValue',
          'setDefaultValues',
          'setErrors',
          'addErrors',
          'clearForm',
          'clearErrors',
          'getFormId',
          'regField',
          'unregField',
          'validateAll',
          'submit',
        ].reduce(reduceMockProps, {}),
      },
      ...[
        'addListItem',
        'removeListItem',
        'reorderListItem',
        'reorderListItemDec',
        'reorderListItemInc',
      ].reduce(reduceMockProps, {}),
    };
    const renderer = createRenderer();
    expect(
      renderer.render(<RepeatableContainer {...props} />)
    ).toMatchSnapshot();
  });
});
