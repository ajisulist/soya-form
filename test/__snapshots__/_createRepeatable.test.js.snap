// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repeatable container should create 3 repeatable components 1`] = `
<div>
  <Repeatable
    addListItem={[Function]}
    form={
      Object {
        "addErrors": [Function],
        "clearErrors": [Function],
        "clearForm": [Function],
        "disable": [Function],
        "disableField": [Function],
        "disableFields": [Function],
        "enable": [Function],
        "enableField": [Function],
        "enableFields": [Function],
        "getFormId": [Function],
        "lockSubmission": [Function],
        "regField": [Function],
        "setDefaultValue": [Function],
        "setDefaultValues": [Function],
        "setErrors": [Function],
        "setValue": [Function],
        "setValues": [Function],
        "submit": [Function],
        "unlockSubmission": [Function],
        "unregField": [Function],
        "validateAll": [Function],
      }
    }
    index={0}
    length={3}
    minLength={0}
    name={[Function]}
    removeListItem={[Function]}
    reorderListItem={[Function]}
    reorderListItemDec={[Function]}
    reorderListItemInc={[Function]}
    totalItem={3}
  />
  <Repeatable
    addListItem={[Function]}
    form={
      Object {
        "addErrors": [Function],
        "clearErrors": [Function],
        "clearForm": [Function],
        "disable": [Function],
        "disableField": [Function],
        "disableFields": [Function],
        "enable": [Function],
        "enableField": [Function],
        "enableFields": [Function],
        "getFormId": [Function],
        "lockSubmission": [Function],
        "regField": [Function],
        "setDefaultValue": [Function],
        "setDefaultValues": [Function],
        "setErrors": [Function],
        "setValue": [Function],
        "setValues": [Function],
        "submit": [Function],
        "unlockSubmission": [Function],
        "unregField": [Function],
        "validateAll": [Function],
      }
    }
    index={1}
    length={3}
    minLength={0}
    name={[Function]}
    removeListItem={[Function]}
    reorderListItem={[Function]}
    reorderListItemDec={[Function]}
    reorderListItemInc={[Function]}
    totalItem={3}
  />
  <Repeatable
    addListItem={[Function]}
    form={
      Object {
        "addErrors": [Function],
        "clearErrors": [Function],
        "clearForm": [Function],
        "disable": [Function],
        "disableField": [Function],
        "disableFields": [Function],
        "enable": [Function],
        "enableField": [Function],
        "enableFields": [Function],
        "getFormId": [Function],
        "lockSubmission": [Function],
        "regField": [Function],
        "setDefaultValue": [Function],
        "setDefaultValues": [Function],
        "setErrors": [Function],
        "setValue": [Function],
        "setValues": [Function],
        "submit": [Function],
        "unlockSubmission": [Function],
        "unregField": [Function],
        "validateAll": [Function],
      }
    }
    index={2}
    length={3}
    minLength={0}
    name={[Function]}
    removeListItem={[Function]}
    reorderListItem={[Function]}
    reorderListItemDec={[Function]}
    reorderListItemInc={[Function]}
    totalItem={3}
  />
</div>
`;
